wasm/libs/default_executor/src/multisig/transaction.rs

```diff
@@ -0,0 +1,249 @@
+//! Validation and execution logic of instructions for multisig transactions
+
+use alloc::{
+    collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+    vec::Vec,
+};
+
+use super::*;
+
+impl VisitExecute for MultisigPropose {
+    fn visit(&self, executor: &mut Executor) {
+        let host = executor.host();
+        let proposer = executor.context().authority.clone();
+        let target_account = self.account.clone();
+        let multisig_role = multisig_role_for(&target_account);
+        let instructions_hash = HashOf::new(&self.instructions);
+
+        let Ok(_role_found) = host
+            .query(FindRolesByAccountId::new(proposer))
+            .filter_with(|role_id| role_id.eq(multisig_role))
+            .execute_single()
+        else {
+            deny!(executor, "not qualified to propose multisig");
+        };
+
+        let Err(_proposal_not_found) = host.query_single(FindAccountMetadata::new(
+            target_account.clone(),
+            approvals_key(&instructions_hash),
+        )) else {
+            deny!(executor, "multisig proposal duplicates")
+        };
+    }
+
+    fn execute(self, executor: &mut Executor) -> Result<(), ValidationFail> {
+        let host = executor.host().clone();
+        let proposer = executor.context().authority.clone();
+        let target_account = self.account;
+        let instructions_hash = HashOf::new(&self.instructions);
+        let signatories: BTreeMap<AccountId, u8> = host
+            .query_single(FindAccountMetadata::new(
+                target_account.clone(),
+                SIGNATORIES.parse().unwrap(),
+            ))
+            .dbg_unwrap()
+            .try_into_any()
+            .dbg_unwrap();
+
+        // Recursively deploy multisig authentication down to the personal leaf signatories
+        for signatory in signatories.keys().cloned() {
+            let is_multisig_again = host
+                .query(FindRoleIds)
+                .filter_with(|role_id| role_id.eq(multisig_role_for(&signatory)))
+                .execute_single_opt()
+                .dbg_unwrap()
+                .is_some();
+
+            if is_multisig_again {
+                let propose_to_approve_me = {
+                    let approve_me =
+                        MultisigApprove::new(target_account.clone(), instructions_hash.clone());
+
+                    MultisigPropose::new(signatory, [approve_me.into()].to_vec())
+                };
+                let init_authority = executor.context().authority.clone();
+                // Authorize as the multisig account
+                executor.context_mut().authority = target_account.clone();
```

Comment:
who is executing here? do they own `target_account`'s private key?

Reply:
Here authority is elevated from a signatory's one to the multisig account's one. Private keys don't matter as authentication already passed